pkg/statistics/handle/autoanalyze/priorityqueue/calculator.go

```diff
@@ -14,23 +14,55 @@
 
 package priorityqueue
 
-// WeightCalculator is an interface for calculating weights of analysis jobs.
-type WeightCalculator interface {
-	CalculateWeight(job *TableAnalysisJob) float64
-}
+import "math"
+
+const (
+	// EventNone represents no special event.
+	eventNone = 0.0
+	// EventNewIndex represents a special event for newly added indexes.
+	eventNewIndex = 2.0
+)
+
+// TODO: make these configurable.
+const (
+	changeRatioWeight = 0.6
+	sizeWeight        = 0.1
+	analysisInterval  = 0.3
+)
 
 // PriorityCalculator implements the WeightCalculator interface.
-type PriorityCalculator struct {
-	threshold float64
-}
+type PriorityCalculator struct{}
 
-// NewPriorityCalculator creates a new PriorityCalculator with the given threshold.
-func NewPriorityCalculator(threshold float64) *PriorityCalculator {
-	return &PriorityCalculator{threshold: threshold}
+// NewPriorityCalculator creates a new PriorityCalculator.
+//
+// For more information, please visit:
+// https://github.com/pingcap/tidb/blob/master/docs/design/2023-11-29-priority-queue-for-auto-analyze.md
+func NewPriorityCalculator() *PriorityCalculator {
+	return &PriorityCalculator{}
 }
 
 // CalculateWeight calculates the weight based on the given rules.
-func (*PriorityCalculator) CalculateWeight(_ *TableAnalysisJob) float64 {
-	// TODO: implement the weight calculation
-	return 1
+// - Table Change Ratio (Change Ratio): Accounts for 60%
+// - Table Size (Size): Accounts for 10%
+// - Analysis Interval (Analysis Interval): Accounts for 30%
+// priority_score calculates the priority score based on the following formula:
+//
+//	priority_score = (0.6 * math.Log10(1 + ChangeRatio) +
+//	                  0.1 * (1 - math.Log10(1 + TableSize)) +
+//	                  0.3 * math.Log10(1 + math.Sqrt(AnalysisInterval)) +
+//	                  special_event[event])
+func (pc *PriorityCalculator) CalculateWeight(job *TableAnalysisJob) float64 {
+	changeRatio := 100 * job.ChangePercentage
```

Comment:
What does this `100 *` mean? Isn't `job.ChangePercentage` already a "ratio"?

Reply:
After normalization, a ratio of 0.1 to 10 yields a better value.
<img width="1755" alt="image" src="https://github.com/pingcap/tidb/assets/29879298/84ff73e6-973a-4b83-a8ff-5ef250a781a1">
vs
<img width="1735" alt="image" src="https://github.com/pingcap/tidb/assets/29879298/4f197874-b230-4fb8-a308-ca64862b9da6">

Reply:
Maybe we will use it in other features or places in the future, so I kept it as the original value.

Reply:
Also, this would align with how the old implementation calculates the change ratio.